src/UnitTest/Core/Tensor.cpp

```diff
@@ -1247,6 +1267,84 @@ TEST_P(TensorPermuteDevices, ReduceSumNotKeepDim) {
+    dst = src.Sum({}, false);
+    EXPECT_EQ(dst.GetShape(), SizeVector({}));
+    EXPECT_EQ(dst.ToFlatVector<float>(), std::vector<float>({1}));
+    dst = src.Sum({}, true);
+    EXPECT_EQ(dst.GetShape(), SizeVector({}));
+    EXPECT_EQ(dst.ToFlatVector<float>(), std::vector<float>({1}));
+
+    // np.sum(np.ones(()), axis=(out_of_bound), keepdims=*)
+    EXPECT_THROW(dst.Sum({0}, false), std::runtime_error);
+    EXPECT_THROW(dst.Sum({-1}, false), std::runtime_error);
+    EXPECT_THROW(dst.Sum({-1}, true), std::runtime_error);
+    EXPECT_THROW(dst.Sum({1}, false), std::runtime_error);
+    EXPECT_THROW(dst.Sum({1}, true), std::runtime_error);
+
+    // Emtpy reduction axis ().
+    // This is different from "reduce all axis" (e.g. np.sum(a)).
+    // np.sum(np.ones((0)), axis=(), keepdims=*)
+    src = Tensor::Ones({0}, Dtype::Float32, device);
+    dst = src.Sum({}, false);
+    EXPECT_EQ(dst.GetShape(), SizeVector({0}));
+    EXPECT_EQ(dst.ToFlatVector<float>(), std::vector<float>({}));
+    dst = src.Sum({}, true);
+    EXPECT_EQ(dst.GetShape(), SizeVector({0}));
+    EXPECT_EQ(dst.ToFlatVector<float>(), std::vector<float>({}));
+
+    // np.sum(np.ones((0)), axis=(0,), keepdims=*), fill with identity
+    src = Tensor::Ones({0}, Dtype::Float32, device);
+    dst = src.Sum({0}, false);
+    EXPECT_EQ(dst.GetShape(), SizeVector({}));  // 1D becomes 0D.
+    EXPECT_EQ(dst.ToFlatVector<float>(), std::vector<float>({0}));
+    dst = src.Sum({0}, true);
+    EXPECT_EQ(dst.GetShape(), SizeVector({1}));  // Remains 1D, but with size 1.
+    EXPECT_EQ(dst.ToFlatVector<float>(), std::vector<float>({0}));
+
+    // np.sum(np.ones((0, 2)), axis=(), keepdims=*)
+    src = Tensor::Ones({0, 2}, Dtype::Float32, device);
+    dst = src.Sum({}, false);
+    EXPECT_EQ(dst.GetShape(), SizeVector({0, 2}));
+    EXPECT_EQ(dst.ToFlatVector<float>(), std::vector<float>({}));
+    dst = src.Sum({}, true);
+    EXPECT_EQ(dst.GetShape(), SizeVector({0, 2}));
+    EXPECT_EQ(dst.ToFlatVector<float>(), std::vector<float>({}));
+
+    // np.sum(np.ones((0, 2)), axis=(0,), keepdims=*), fill with identity
+    src = Tensor::Ones({0, 2}, Dtype::Float32, device);
+    dst = src.Sum({0}, false);
+    EXPECT_EQ(dst.GetShape(), SizeVector({2}));
+    EXPECT_EQ(dst.ToFlatVector<float>(), std::vector<float>({0, 0}));
+    dst = src.Sum({0}, true);
+    EXPECT_EQ(dst.GetShape(), SizeVector({1, 2}));
+    EXPECT_EQ(dst.ToFlatVector<float>(), std::vector<float>({0, 0}));
+
+    // np.sum(np.ones((0, 2)), axis=(1,), keepdims=*)
+    src = Tensor::Ones({0, 2}, Dtype::Float32, device);
+    dst = src.Sum({1}, false);
+    EXPECT_EQ(dst.GetShape(), SizeVector({0}));
+    EXPECT_EQ(dst.ToFlatVector<float>(), std::vector<float>({}));
+    dst = src.Sum({1}, true);
+    EXPECT_EQ(dst.GetShape(), SizeVector({0, 1}));
+    EXPECT_EQ(dst.ToFlatVector<float>(), std::vector<float>({}));
+
+    // np.sum(np.ones((0, 2)), axis=(0, 1), keepdims=*), fill with identity
```

Comment:
can we add (some of) these to test_core.py (and compare to numpy)?

Reply:
done, added in `test_core.py::test_reduction_special_shapes`